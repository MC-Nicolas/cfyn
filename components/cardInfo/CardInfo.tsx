import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { selectAreDataVisible } from '../../redux/data/data.selectors';

import styles from '../../styles/CardInfoBar.module.scss';
import TrendingUpIcon from '@material-ui/icons/TrendingUp';
import TrendingDownIcon from '@material-ui/icons/TrendingDown';
import TrendingFlatIcon from '@material-ui/icons/TrendingFlat';

interface Props {
  title: string;
  isTrendUp: boolean;
  isTrendFlat?: boolean;
  amount: number;
  areDataVisible: boolean;
}

const CardInfo = ({
  title,
  isTrendUp,
  amount,
  areDataVisible,
  isTrendFlat = false,
}: Props) => {
  return (
    <div className={styles.cardInfo}>
      <div className={styles.leftContainer}>
        <p data-testid='CardInfoTitle' className={styles.leftContainerTitle}>
          {title.replace(/^\w/, (c) => c.toUpperCase())}
        </p>
        <p
          data-testid='cardInfoAmountIndicator'
          className={`${styles.leftContainerAmount} ${
            areDataVisible
              ? isTrendFlat
                ? 'colorOrange'
                : isTrendUp
                ? 'colorGreen'
                : 'colorRed'
              : ''
          }`}
        >
          {areDataVisible ? (isTrendUp ? `+${amount}` : `-${amount}`) : '...'}â‚¬
        </p>
      </div>
      <div className={styles.rightContainer}>
        {areDataVisible && isTrendUp && !isTrendFlat && (
          <TrendingUpIcon className={styles.rightContainerIconUp} />
        )}
        {areDataVisible && !isTrendUp && !isTrendFlat && (
          <TrendingDownIcon className={styles.rightContainerIconDown} />
        )}
        {areDataVisible && isTrendFlat && (
          <TrendingFlatIcon className={styles.rightContainerIconFlat} />
        )}
      </div>
    </div>
  );
};

const mapStateToProps = createStructuredSelector({
  areDataVisible: selectAreDataVisible,
});

export default connect(mapStateToProps)(CardInfo);
