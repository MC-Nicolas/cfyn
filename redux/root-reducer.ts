import { combineReducers } from 'redux';

import UserReducer from './user/user.reducer';
import DataReducer from './data/data.reducer';
import DatabasesReducer from './databases/databases.reducer';
import LoaderReducer from './loader/loader.reducer';
import PagesReducer from './pages/pages.reducer';

const rootReducer = combineReducers({
  user: UserReducer,
  data: DataReducer,
  databases: DatabasesReducer,
  loader: LoaderReducer,
  pages: PagesReducer,
});

export default rootReducer;
