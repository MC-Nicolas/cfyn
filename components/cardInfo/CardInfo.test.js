/**
 * @jest-environment jsdom
 */

import React from 'react';
import {
  render,
  cleanup,
  fireEvent,
  getNodeText,
} from '@testing-library/react';
import { store } from '../../redux/store';
import { Provider } from 'react-redux';
import CardInfo from './CardInfo';

afterEach(cleanup);

describe('<CardInfo />', () => {
  it('Positive amount', () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <CardInfo title='total' isTrendUp={true} amount={20} />
      </Provider>
    );
    const amountIndicator = getByTestId('cardInfoAmountIndicator');
    const cardTitle = getByTestId('CardInfoTitle');
    expect(cardTitle.textContent).toBe('Total');
    expect(amountIndicator.textContent).toBe('+20€');
    expect(amountIndicator).toHaveClass('colorGreen');
    expect(amountIndicator).toMatchSnapshot();
  });
  it('Negative Amount', () => {
    const { getByTestId, debug } = render(
      <Provider store={store}>
        <CardInfo title='income 1' isTrendUp={false} amount={-100} />
      </Provider>
    );
    const amountIndicator = getByTestId('cardInfoAmountIndicator');
    const cardTitle = getByTestId('CardInfoTitle');
    expect(cardTitle.textContent).toBe('Income 1');
    expect(amountIndicator.textContent).toBe('-100€');
    expect(amountIndicator).toHaveClass('colorRed');
    expect(amountIndicator).toMatchSnapshot();
  });

  it('Equal amount', () => {
    const { getByTestId, debug } = render(
      <Provider store={store}>
        <CardInfo
          title='balance'
          isTrendUp={false}
          isTrendFlat={true}
          amount={0}
        />
      </Provider>
    );
    const amountIndicator = getByTestId('cardInfoAmountIndicator');
    const cardTitle = getByTestId('CardInfoTitle');
    expect(cardTitle.textContent).toBe('Balance');
    expect(amountIndicator.textContent).toBe('0€');
    expect(amountIndicator).toHaveClass('colorOrange');
    expect(amountIndicator).toMatchSnapshot();
  });
});
