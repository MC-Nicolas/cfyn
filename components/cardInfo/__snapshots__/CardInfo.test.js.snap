// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardInfo /> Equal amount 1`] = `
<p
  class="undefined colorOrange"
  data-testid="cardInfoAmountIndicator"
>
  0
  €
</p>
`;

exports[`<CardInfo /> Negative Amount 1`] = `
<p
  class="undefined colorRed"
  data-testid="cardInfoAmountIndicator"
>
  -100
  €
</p>
`;

exports[`<CardInfo /> Positive amount 1`] = `
<p
  class="undefined colorGreen"
  data-testid="cardInfoAmountIndicator"
>
  +20
  €
</p>
`;
